browsers/private/update-tool/index.mts

```diff
@@ -6,64 +6,167 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {resolveBuildId, Browser, BrowserPlatform, BrowserTag} from '@puppeteer/browsers';
-import {mkdtemp} from 'node:fs/promises';
-import os from 'node:os';
-import path from 'node:path';
-import {downloadAndHashBinariesForBrowser} from './download.mjs';
-import {generateRepositorySetupBzlFile} from './generation.mjs';
-import fs from 'node:fs/promises';
-
-main().catch(e => {
+import { resolveBuildId, Browser, BrowserPlatform } from "@puppeteer/browsers";
+import { mkdtemp } from "node:fs/promises";
+import os from "node:os";
+import path from "node:path";
+import { downloadAndHashBinariesForBrowser } from "./download.mjs";
+import {
+  generateRepoInfo,
+  generateVersionsBzlFile,
+  Versions,
+} from "./generation.mjs";
+import fs from "node:fs/promises";
+import { getChromeMilestones, getFirefoxMilestones } from "./versions.mjs";
+
+main().catch((e) => {
   console.error(e);
   process.exitCode = 1;
 });
 
-async function main() {
-  const tmpDir = await mkdtemp(path.join(os.tmpdir(), 'rules_browsers_tmp-'));
-
-  const [chromeBuildId, firefoxBuildId, chromedriverBuildId] = await Promise.all([
-    resolveBuildId(Browser.CHROMEHEADLESSSHELL, null!, BrowserTag.STABLE),
-    // TODO: Explicitly forced to 135 as the latest stable seems to cause instability in RBE.
-    // See: https://github.com/devversion/rules_browsers/commit/afa52d3dffa7b41e8049821a3d859524cf01597b.
-    resolveBuildId(Browser.FIREFOX, null!, 'stable_135.0'),
-    resolveBuildId(Browser.CHROMEDRIVER, null!, BrowserTag.STABLE),
-  ]);
+interface WriteVersionsOptions {
+  browser: Browser;
+  milestones: string[];
+  tmpDir: string;
+  workspaceRoot: string;
+  excludeFilesForPerformance?: Partial<Record<BrowserPlatform, string[]>>;
+  buildIdToVersion?: (buildId: string) => string;
+}
 
-  const [chromeBinaries, chromedriverBinaries, firefoxBinaries] = await Promise.all([
-    downloadAndHashBinariesForBrowser(
-      tmpDir,
-      // Puppeteer team suggests headless shell as it's more lightweight and faster.
-      Browser.CHROMEHEADLESSSHELL,
-      chromeBuildId,
-      {},
-      {
-        // Exclude log files that Chrome might write to— causing remote cache misses.
-        [BrowserPlatform.LINUX]: ['**/*.log'],
-        [BrowserPlatform.MAC]: ['**/*.log'],
-        [BrowserPlatform.MAC_ARM]: ['**/*.log'],
-        [BrowserPlatform.WIN64]: ['**/*.log'],
-      }
-    ),
-    downloadAndHashBinariesForBrowser(tmpDir, Browser.CHROMEDRIVER, chromedriverBuildId),
-    downloadAndHashBinariesForBrowser(tmpDir, Browser.FIREFOX, firefoxBuildId),
-  ]);
+// We treat chrome-headless-shell as Chromium in the module. This function
+// ensures browser name is written to files as expected.
+function getReadableBrowserName(browser: Browser): string {
+  return browser === Browser.CHROMEHEADLESSSHELL ? "chromium" : browser;
+}
+
+async function writeVersions({
```

Comment:
Maybe there is a better name for this? e.g. `downloadMilestonesAndWriteVersionsBzl`?